src/pml_vqvae/train.py

```diff
@@ -1,17 +1,96 @@
 "Python script to train different models"
 
 import argparse
+from pml_vqvae.baseline.autoencoder import BaselineAutoencoder
+from pml_vqvae.dataset.dataloader import load_data
+import torch
+from torchvision.transforms import RandomCrop
+
+# import wandb
+
+# Hyperparameters
+MODEL = BaselineAutoencoder()
+DATASET = "cifar"
+
+EPOCHS = 100
+LEARNING_RATE = 0.01
+MOMENTUM = 0.9
+N_TRAIN = 1000
+N_TEST = 1000
+
+SEED = 2024
+
+
+# From wandb tutotial https://docs.wandb.ai/tutorials/pytorch
+def train_log(loss, epoch, epochs):
+    # Where the magic happens
+    # wandb.log({"epoch": epoch, "loss": loss}, step=epoch)
+    print(f"Epoch {epoch+1}/{epochs} - Loss: {loss}")
+
+
+def train(model: torch.nn.Module = MODEL, dataset: str = DATASET, epochs: int = EPOCHS):
+    print(f"Training {model} on {dataset} for {epochs} epochs")
+
+    # Load data
+    print("Loading dataset")
+    train_loader, test_loader = load_data(
+        dataset,
+        RandomCrop(128, pad_if_needed=True),
+        n_train=N_TRAIN,
+        n_test=N_TEST,
+        seed=SEED,
+    )
+
+    # wandb.init(
+    #     project="PML-VQVAE",
+    #     config={
+    #         "learning_rate": LEARNING_RATE,
+    #         "architecture": model.name(),
+    #         "dataset": dataset,
+    #         "n_train": N_TRAIN,
+    #         "n_test": N_TEST,
+    #         "epochs": epochs,
+    #     },
+    # )
+
+    loss_fn = torch.nn.MSELoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=LEARNING_RATE, momentum=MOMENTUM)
+
+    # Watch model with wandb
+    # wandb.watch(model, loss_fn, log="all", log_freq=10)
+
+    # Train model
+    losses = []  # losses over epochs
+    vals = []
+    for i in range(epochs):
+        batch_losses = []
+        # iterate over batches
+        for batch_img, _ in train_loader:
+
+            optimizer.zero_grad()
+
+            output = model(batch_img)
+            loss = loss_fn(output, batch_img)
+            batch_losses.append(loss.item())
+
+            loss.backward()
+            optimizer.step()
+
+        epoch_loss = sum(batch_losses) / len(batch_losses)
+        losses.append(epoch_loss)
+        train_log(epoch_loss, i, epochs)
```

Comment:
I think its probably better to write a training function for each model in its own file (e.g. this function could be moved to /basline/autoencoder.py and then this file (train.py) is simply a wrapper for all of thos functions. Reason for that is that we will have different losses, especially for the VQVAE and VAE since they also have to get the latent variable Z, furthermore the pixelcnn will also not be trainable with this function.

Reply:
I prefer just one train function where where we handle the logging and plotting stuff. Well, we can right the vae first into an torch.nn.module, put the encoder -> sample -> decoder logic into the forward function and the loss can be a staticmethod of that model class. 
That was at least my idea to have them all as modules.

Reply:
But if we have just this 1 global train function, how will we implement the copying of the gradients of the codes onto the output of the encoder, furthermore we will have to implement the "stop gradient" ($\text{sg}\[\cdot\]$) operation from the paper, which doesnt follow the usual optimizer.zero_grad() -> loss.backward() -> optimizer.step() logic. At least I think it isnt possible that way.

Still its probably useful to have a function for training the simpler models, in which case I would propose the loss_fn, optimizer and dataloaders be direct inputs into that function